### PR TITLE
Iaas - VM deployment improvements

### LINKED ISSUE #121: New-CmAzIaasVm doesn't allow you to specify the disk caching policy
managed disks on virtual machines have various settings for disk caching which are not exposed in the cmdlet.

### LINKED ISSUE #215: New-CmAzIaasVm add support for availability zones
Add support for availability zones in VMs the New-CmAzIaasVm cmdlet. In the [virtualmachines configuration section](https://docs.trustmarque.com/cloudmarque/reference/commands/new-cmaziaasvm.html) there should be an additional `zone` setting:

``` yaml
    virtualMachines:     # Mandatory
      - name: "string"    # Mandatory
        size: "string"    # Mandatory
        zone: 1         # Optional
        updateGroup: "Critical"    # Optional
        updateFrequency: "Monthly"    # Optional
```

The setting should be:
 - optional
 - numeric
 - ranged between 1 and 3
 - default to 1
 - example should be set to 1

Example: https://github.com/Azure/azure-quickstart-templates/tree/master/101-vm-simple-zones

### LINKED ISSUE #216: New-CmAzIaasVm : move $virtualMachine.fullName Get-CmAzResourceName order
move $virtualMachine.fullName 1st so that nic and os disk match fullname (fullname was last so nic os didnt have 'vm' prefix)

### LINKED ISSUE #226: New-CmAzIaasVm add support for availability sets
Support for availability sets within the New-CmAzIaasVm cmdlet.

Example: https://github.com/Azure/azure-quickstart-templates/blob/master/201-vm-copy-index-loops/azuredeploy.json

Arm template ref: https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2018-10-01/availabilitysets

```yaml
availabilitySets:
- name: 
  location:
  sku:
    name:
    tier:                                            # Standard or Basic
    capacity:
  platformUpdateDomainCount: 
  platformFaultDomainCount:
  proximityPlacementGroup:
  service:
    publish:
      availabilitySet:
```
Add "AvailabilitySet" property to each vm settingsObject. the value should be whats passed into Get-CmAzResourceName to generate the availability set name.

### PR DESCRIPTION
**Pull request by deepend-dev**
- Resolves #121 by implementing caching policy.
- Resolves #215 by implementing support for availability zones.
- Resolves #226 by implementing support for availability sets. 
- Resolves #216

### CODE REVIEW

File: Cloudmarque.Azure/Resources/Project/_names/regions.yml [hunk 1]
```diff
@@ -17,3 +17,5 @@ regions:
   shortname: ukw
 - name: UKWest
   shortname: ukw
+- name: East US
+  shortname: use
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/_names/regions.yml:21]: Should be eus

File: Cloudmarque.Azure/Resources/Project/_names/resources.yml [hunk 2]
```diff
@@ -106,7 +106,7 @@ resources:
   shortname: vm
   minLength: 1
   maxLength: 64
-  generator: VirtualMachine
+  generator: virtualmachine
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/_names/resources.yml:109]: Why the change here?

File: Cloudmarque.Azure/Resources/Project/virtualmachines.yml [hunk 3]
```diff
@@ -1,38 +1,63 @@
...
+          publish:
+            proximityPlacementGroup:
+
+    availabilitySets:
+      - name: set1
+        location: UK South
+        sku:
+          name: Aligned                                          # Aligned or Classic
+          tier:                                                  # Standard or Basic
+          capacity:
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/virtualmachines.yml:41]: Add a comment for capacity

File: azure-integration-pipelines.yml [hunk 4]
```diff
@@ -132,8 +132,10 @@ stages:
 
           Set-CmAzIaasUpdateManagement -SettingsFile "$(Build.SourcesDirectory)/$(moduleName)/Resources/Project/updateManagement.yml" -Verbose
 
+- stage: 'Iaas_Bastion'
```
Comment (tm-luke) [inline azure-integration-pipelines.yml:135]: Do we need an additional step in the pipeline?

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 5]
```diff
@@ -99,6 +99,18 @@
                             "type": ["string", "null"],
                             "title": "NIC",
                             "description": "Global default value to publish on deployed resource type."
+                        },
+                        "proximityPlacementGroup": {
+                            "$id": "#/service/publish/proximityPlacementGroup",
+                            "type": ["string", "null"],
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:105]: Is this completely optional?

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 6]
```diff
@@ -99,6 +99,18 @@
...
+                            "title": "proximityPlacementGroup",
+                            "description": "Global default value to publish on deployed resource type."
+                        },
+                        "availabilitySet": {
+                            "$id": "#/service/publish/availabilitySet",
+                            "type": ["string", "null"],
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:111]: Is this completely optional?

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 7]
```diff
@@ -165,6 +177,174 @@
...
+                                "description": "Contains list of details required to deploy proximityPlacementGroups.",
+                                "items": {
+                                    "$id": "#/groups/items/allOf/proximityPlacementGroups/items",
+                                    "minLength" : 1,
+                                    "allOf": [
+                                        {
+                                            "$id": "#/groups/items/allOf/proximityPlacementGroups/items/allOf",
+                                            "type": "object",
+                                            "required": [
+                                                "name"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:193]: If we do define a proximityPlacementGroup, is name the only propety that is strictly required?

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 8]
```diff
@@ -165,6 +177,174 @@
...
+                                                    "$id": "#/groups/items/allOf/proximityPlacementGroups/items/allOf/name",
+                                                    "type": "string",
+                                                    "title": "Name",
+                                                    "description": "Becomes part of the generated proximityPlacementGroups name."
+                                                },
+                                                "location": {
+                                                    "$id": "#/groups/items/allOf/proximityPlacementGroups/items/allOf/location",
+                                                    "type": ["string", "null"],
+                                                    "title": "Location",
+                                                    "description": "proximityPlacementGroups location"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:206]: ProximityPlacementGroup's location.

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 9]
```diff
@@ -257,6 +437,19 @@
                                                     "title": "Location",
                                                     "description": "Virtual machine deployment location"
                                                 },
+                                                "zone": {
+                                                    "$id": "#/groups/items/allOf/virtualMachines/items/allOf/zone",
+                                                    "type": ["integer", "null"],
+                                                    "title": "zone",
+                                                    "description": "Virtual machine deployment zone.",
+                                                    "enum": [ 1, 2, 3, ""]
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:445]: Should be null instead of ""

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 10]
```diff
@@ -165,6 +177,174 @@
...
+                                                    "$id": "#/groups/items/allOf/availabilitySets/items/allOf/location",
+                                                    "type": ["string", "null"],
+                                                    "title": "Location",
+                                                    "description": "availabilitySets location"
+                                                },
+                                                "platformUpdateDomainCount": {
+                                                    "$id": "#/groups/items/allOf/availabilitySets/items/allOf/platformUpdateDomainCount",
+                                                    "type": ["integer", "null"],
+                                                    "title": "platformUpdateDomainCount",
+                                                    "description": "Update Domain count."
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:273]: Can this be elaborated on more?

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 11]
```diff
@@ -165,6 +177,174 @@
...
+                                                },
+                                                "platformFaultDomainCount": {
+                                                    "$id": "#/groups/items/allOf/availabilitySets/items/allOf/platformFaultDomainCount",
+                                                    "type": ["integer", "null"],
+                                                    "title": "platformFaultDomainCount",
+                                                    "description": "Fault Domain count."
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:279]: Can this be elaborated on more?

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 12]
```diff
@@ -165,6 +177,174 @@
...
+                                                    "required": [
+                                                        "name"
+                                                    ],
+                                                    "properties": {
+                                                        "name": {
+                                                            "$id": "#/groups/items/allOf/availabilitySets/items/allOf/sku/name",
+                                                            "type": "string",
+                                                            "title": "name",
+                                                            "description": "The sku name.",
+                                                            "enum": [ "Classic", "classic", "Aligned", "aligned"]
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:301]: For enums, just use one set of values for now actually, i will raise a separate issue for handling enum casing

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 13]
```diff
@@ -310,22 +506,51 @@
                                                             "title": "ExpandableDiskSizeGB",
                                                             "description": "The size of the disk in gigabytes. (Overwrites disk size in vm image).",
                                                             "minimum": 1
+                                                        },
+                                                        "caching": {
+                                                            "$id": "#/groups/items/allOf/virtualMachines/items/allOf/osDisk/caching",
+                                                            "type": "string",
+                                                            "title": "caching",
+                                                            "description": "Defines disk caching strategy",
+                                                            "enum": ["none", "None", "ReadOnly", "Readonly", "readOnly", "readonly", "ReadWrite", "readwrite", "readWrite", "Readwrite"]
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:515]: For enums, just use one set of values for now actually, i will raise a separate issue for handling enum casing

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 14]
```diff
@@ -310,22 +506,51 @@
...
+                                                                        "title": "diskSizeGB",
+                                                                        "description": "The size of the disk in gigabytes.",
+                                                                        "minimum": 1
+                                                                    },
+                                                                    "caching": {
+                                                                        "$id": "#/groups/items/allOf/virtualMachines/items/allOf/dataDisks/items/allOf/caching",
+                                                                        "type": "string",
+                                                                        "title": "Caching",
+                                                                        "description": "Defines disk caching strategy",
+                                                                        "enum": ["none", "None", "ReadOnly", "Readonly", "readOnly", "readonly", "ReadWrite", "readwrite", "readWrite", "Readwrite"]
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:550]: For enums, just use one set of values for now actually, i will raise a separate issue for handling enum casing

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json [hunk 15]
```diff
@@ -207,11 +252,13 @@
                             "dependsOn": [
                                 "[parameters('VM').nicName]"
                             ],
+                            "zones": "[if(equals(parameters('VM').zone, 'None'), json('null'), array(parameters('VM').zone))]",
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json:255]: done

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 16]
```diff
@@ -125,8 +125,76 @@
...
+
+						$placementGroup.generatedName = Get-CmAzResourceName -Resource "ProximityPlacementGroup" -Architecture "IaaS" -Region $placementGroup.location -Name $placementGroup.name
+						Set-GlobalServiceValues -GlobalServiceContainer $SettingsObject -ServiceKey "ProximityPlacementGroup" -ResourceServiceContainer $placementGroup
+					}
+				}
+				else {
+					$resourceGroup.proximityPlacementGroups = @(
+						@{
+							name = ""
+							generatedName = "none";
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:141]: it is required for search we do later..

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 17]
```diff
@@ -125,8 +125,76 @@
...
+								}
+							}
+						}
+					)
+				}
+
+				if ($resourceGroup.availabilitySets) {
+					foreach ($set in $resourceGroup.availabilitySets) {
+
+						$set.generatedName = Get-CmAzResourceName -Resource "AvailabilitySet" -Architecture "IaaS" -Region $set.location -Name $set.name
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:155]: required for search or we need to call resourcename function repeatedly

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 18]
```diff
@@ -125,8 +125,76 @@
...
+
+						if ($set.proximityPlacementGroup) {
+
+							$set.proximityPlacementGroup = ($resourceGroup.proximityPlacementGroups | Where-Object {$_.name -eq $set.proximityPlacementGroup}).generatedName
+
+							if(!$set.proximityPlacementGroup){
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:161]: yes or we get no object which throws an error in arm template

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 19]
```diff
@@ -160,25 +228,47 @@
 					$virtualMachine.osDisk.Name = Get-CmAzResourceName -Resource "OSDisk" -Architecture "IaaS" -Region $virtualMachine.location -Name $virtualMachine.name
 					$virtualMachine.fullName = Get-CmAzResourceName -Resource "VirtualMachine" -Architecture "IaaS" -Region $virtualMachine.location -Name $virtualMachine.name
 
+					$virtualMachine.osDisk.caching ??= "None"
+
+					$virtualMachine.zone ??= "None"
+
+					if ($virtualMachine.availabilitySet){
+
+						$virtualMachine.availabilitySet = ($resourceGroup.availabilitySets | Where-Object {$_.name -eq $virtualMachine.availabilitySet}).generatedName
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:237]: it's for comparison of name provided in settings file..

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 20]
```diff
@@ -160,25 +228,47 @@
...
+
+						if(!$virtualMachine.availabilitySet){
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:239]: done

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 21]
```diff
@@ -160,25 +228,47 @@
...
 
-					for ($i = 0; $i -lt $virtualMachine.dataDiskSizes.count; $i++) {
+					if ($virtualMachine.dataDisks) {
 
-						$virtualMachine.dataDisks += @{
-							"Name"         = Get-CmAzResourceName -Resource "DataDisk" -Architecture "IaaS" -Region $virtualMachine.location -Name "$($virtualMachine.name)$($i + 1)";
-							"Lun"          = $i + 1;
-							"CreateOption" = "Empty";
-							"DiskSizeGB"   = $virtualMachine.dataDiskSizes[$i]
+						foreach ($disk in $virtualMachine.dataDisks) {
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:252]: done

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 22]
```diff
@@ -160,25 +228,47 @@
...
+
+							$disk.name = Get-CmAzResourceName -Resource "DataDisk" -Architecture "IaaS" -Region $virtualMachine.location -Name "$($virtualMachine.name)$($i + 1)";
+							$disk.Lun = $i + 1;
+							$disk.CreateOption = "Empty";
+							$disk.caching ??= "None"
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:257]: done

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json (hunk 5 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:105]: yes

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json (hunk 6 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:111]: yes

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json (hunk 7 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:193]: yes

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json (hunk 8 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:206]: done

File: Cloudmarque.Azure/Resources/Project/_names/resources.yml (hunk 2 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/_names/resources.yml:109]: generators are all small case.. for consistency

File: Cloudmarque.Azure/Resources/Project/virtualmachines.yml (hunk 3 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/virtualmachines.yml:41]: its an integer

File: azure-integration-pipelines.yml (hunk 4 above)
Comment (deepend-dev) [reply azure-integration-pipelines.yml:135]: it is to seperate out bastion hosts from rest of infra.. because it takes 45 mins to deploy.. which is very annoying if you want to do just staged deployment of vm or networking only ,,

Comment (tm-luke) [reply azure-integration-pipelines.yml:135]: This job should be running in parallel with everything else anyway, plus this pipeline ideally is to serve as a smoke test for Cloudmarque, not for testing a singular aspect.

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 18 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: Add space before {

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 23]
```diff
@@ -160,25 +228,47 @@
 					$virtualMachine.osDisk.Name = Get-CmAzResourceName -Resource "OSDisk" -Architecture "IaaS" -Region $virtualMachine.location -Name $virtualMachine.name
 					$virtualMachine.fullName = Get-CmAzResourceName -Resource "VirtualMachine" -Architecture "IaaS" -Region $virtualMachine.location -Name $virtualMachine.name
 
+					$virtualMachine.osDisk.caching ??= "None"
```
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: Ah realised this is an azure supported value not a default value with use in code, can you revert to None?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 24]
```diff
@@ -160,25 +228,47 @@
...
+
+					$virtualMachine.zone ??= "None"
```
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: Ah realised this is an azure supported value not a default value with use in code, can you revert to None?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 22 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: Ah realised this is an azure supported value not a default value with use in code, can you revert to None?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 19 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: Can you do comparison before the name generation?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 17 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:155]: Can you do comparison before the name generation?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 16 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:141]: Can you do comparison before the name generation?

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 23 above)
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: ok

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 19 above)
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1]: nope.. because actually need generatedname for arm template to work... we compare name and fetch generatedname and pass it to arm template

**Review by tm-luke — APPROVED** [review]
